software/tracksight/frontend/lib/contexts/DashboardLayout.tsx

```diff
@@ -0,0 +1,135 @@
+"use client";
+
+import type { Widget, WIDGET_TYPE } from "@/lib/types/Widget";
+import React, { createContext, useContext, useState } from "react";
+import { IS_DEBUG } from "@/lib/constants";
+
+type DashboardLayoutType = {
+  // NOTE(evan): Might cause performance issues with keeping widget state together with
+  //             card state, if it becomes an issue move to per-card widget state.
+  widgets: Widget<WIDGET_TYPE>[];
+
+  addWidget: (newWidget: Widget<WIDGET_TYPE>) => void;
+  removeWidget: (widgetToRemove: string) => void;
+  editWidget: (
+    widgetToEdit: string,
+    newWidgetData: Partial<Widget<WIDGET_TYPE>>
+  ) => void;
+};
+
+const DashboardLayoutContext = createContext<DashboardLayoutType | undefined>(undefined);
+
+const DashboardLayoutProvider = ({ children }: { children: React.ReactNode }) => {
+  const [widgets, setWidgets] = useState<Widget<WIDGET_TYPE>[]>([
+    // TODO(evan): Remove this it's just temporary data, should be pulled from local storage
+    //             or something similar
```

Comment:
i agree, I know there's a few nice localstorage hooks online, maybe you can make one youself if ur feeling it